File: src/stat/mkgraph.pl
#!/usr/bin/perl
# $ProjectId$
#
# $Id: mkgraph.pl 1.1 Fri, 21 Nov 1997 17:06:57 +0000 andrej $
#
# Nari¹i grafe:
# pie chart porazdelitev cen
# gibanje poravnalnih cen do sedaj
# gibanje porazdelitev poravnalnih cen do sedaj

use GIFgraph::pie;
use GIFgraph::lines;
use Pg;
#use getopt;

# nastavitve:
$convertor = "/net/rozle/export/share/sol/bin/convert"; 
#$destdir = "/opt/borza/HTML/stat/";
$destdir = "./";
#$dbname = "volitve";

$conn = Pg::connectdb("dbname = volitve");

# Popravi datum:
sub fixdate {
    return substr($_[0], 3, 2) . "." . substr($_[0], 0, 2);
}

# Izraèunaj pie graph:
sub Pie {
Pg::doQuery($conn, "select papir_id, tecaj from tecajnica where datum='yesterday'", \@db_data);
  
  $sum = 0;

  for $i ( 0 .. $#db_data ) {
      $sum += $db_data[$i][1];
  }

  @data = ([],[]);

  for $i ( 0 .. $#db_data ) {
      push(@{@data[0]}, sprintf("%s: %5.2f", unpack("A8",pack("a8",$db_data[$i][0])), ($db_data[$i][1]/$sum) * 100));
      push(@{@data[1]}, ($db_data[$i][1]/$sum)*100);
  }

  $pie_graph = new GIFgraph::pie( );
  $pie_graph->set( 'title' => 'Porazdelitev poravnalnih cen',
#                 'label' => 'Datum',
		  'axislabelclr' => 'black' );
#  print "About to plot\n";
  $pie_graph->plot_to_gif( "${destdir}napoved.gif", \@data );
  system("$convertor -geometry 100x100 -colors 16 ${destdir}napoved.gif ${destdir}napoved-chip.gif");
}

sub Line {
# Nari¹i line grafe:

  Pg::doQuery($conn, "select datum, sum(tecaj) from tecajnica where datum<'today' group by datum order by datum", \@db_sum);

  Pg::doQuery($conn, "select papir_id from papirji order by papir_id", \@db_papirji);

    @cene_data = ([]);
    @przd_data = ([]);
    @sum_data = ([],[]);
    %papirji = ();

    for $i ( 0 .. $#db_papirji ) {
 	push(@cene_data, []);
 	push(@przd_data, []);
	$papirji{$db_papirji[$i][0]} = $i+1;
    }

    for $i ( 0 .. $#db_sum) {
 	$datum = &fixdate($db_sum[$i][0]);
	$db_datum = $db_sum[$i][0];
	$sum = $db_sum[$i][1];
 	push(@{@cene_data[0]}, $datum);
 	push(@{@przd_data[0]}, $datum);
	push(@{@sum_data[0]}, $datum);
	push(@{@sum_data[1]}, $sum);
      Pg::doQuery($conn, "select papir_id, tecaj from tecajnica where datum='$db_datum'", \@db_data);	
 	for $j ( 0 .. $#{@db_papirji}) {
	    push(@{@cene_data[$j+1]}, 0);
	    push(@{@przd_data[$j+1]}, 0);
	}
	for $j ( 0 .. $#{@db_data}) {
	    $cene_data[$papirji{$db_data[$j][0]}][$i] = $db_data[$j][1];
	    $przd_data[$papirji{$db_data[$j][0]}][$i] = $db_data[$j][1]/$sum;
	}
    }
    
    $cene_graph = new GIFgraph::lines(600, 400);
    $cene_graph->set( 'title' => 'Poravnalne cene',
#                 'label' => 'Datum',
		     'axislabelclr' => 'black',
		     'line_width' => 10);
    $cene_graph->plot_to_gif( "${destdir}cene.gif", \@cene_data );
    system("$convertor -geometry 100x100 -colors 16 ${destdir}cene.gif ${destdir}cene-chip.gif");

    $przd_graph = new GIFgraph::lines(600, 400);
    $przd_graph->set( 'title' => 'Porazdelitev poravnalnih cen',
		     'axislabelclr' => 'black',
		     'line_width' => 10);
    $przd_graph->plot_to_gif( "${destdir}porazd.gif", \@przd_data );
    system("$convertor -geometry 100x100 -colors 16 ${destdir}porazd.gif ${destdir}porazd-chip.gif");

    $vsota_graph = new GIFgraph::lines(600, 400);
    $vsota_graph->set( 'title' => 'Vsota poravnalnih cen',
		     'axislabelclr' => 'black',
		     'line_width' => 10 );
    $vsota_graph->plot_to_gif( "${destdir}vsota.gif", \@sum_data );
    system("$convertor -geometry 100x100 -colors 16 ${destdir}vsota.gif ${destdir}vsota-chip.gif");

}

Pie;
#Line;

exit;

 

File: src/stat/mkgraph.py
#!/net/rozle/export/share/sol/bin/python

import pg95
import os
import string
import copy

pg95.set_defbase("volitve");
conn = pg95.connect();
grbatch="/net/rozle/export/share/sol/bin/grbatch"
tmpdir="tmp/"
outdir="/opt/borza/HTML/stat/"
#outdir="tmp/"
converter="/net/rozle/export/share/sol/bin/convert"
pardir="/opt/borza/wwwstat/"

def FixDate(date):
    return  date + (4712 + 1997)*365 + 1500 + 480 - 9
#    return date[3:5] + '.' + date[0:2]


def MkGraph(data, name, flags=""):
    f = open(tmpdir + name + ".dat", "w")
    f.write(data)
    f.close()
    os.system("%(grbatch)s -p %(pardir)s%(name)s.par -device 1 -eps -noask -printfile %(tmpdir)s%(name)s.eps %(flags)s %(tmpdir)s%(name)s.dat" % {
	'grbatch': grbatch,
	'name': name,
	'tmpdir': tmpdir,
	'flags': flags,
	'pardir': pardir})
    os.system("%(converter)s -rotate 90 %(tmpdir)s%(name)s.eps %(outdir)s%(name)s.gif" % {
	'converter': converter,
	'tmpdir': tmpdir,
	'name': name,
	'outdir': outdir})
    os.system("%(converter)s -geometry 100x100 -colors 16 %(outdir)s%(name)s.gif %(outdir)s%(name)s-chip.gif" % { 
	'converter': converter,
	'outdir': outdir,
	'name': name})

# vsota vseh poravnalnih cen:
# tu ugotovimo tudi ¹tevilo dnevov:
db_vsota = conn.query("select datum, sum(tecaj) from tecajnica where datum<'today' group by datum order by datum")

empty = []
for d in db_vsota:
    empty.append(0)

# Papirji:
db_papirji = conn.query("select papir_id from papirji order by papir_id");
data_cene = {}
for p in db_papirji:
    data_cene[p[0]] = copy.copy(empty)


str_vsota = ""
data_vsota = {}
datumi = []

for i in db_vsota:
    #    s = %s % FixDate(i[0])
    datum = "%d" % FixDate(db_vsota.index(i))
    datumi.append(i[0])
    data_vsota[i[0]]= string.atof(i[1])
    str_vsota = str_vsota + datum
    str_vsota = str_vsota + " " + i[1]
    str_vsota = str_vsota + "\n"    

db_tecaji = conn.query("select datum, papir_id, tecaj from tecajnica where datum<'today' order by datum, papir_id")

for t in db_tecaji:
    data_cene[t[1]][datumi.index(t[0])] = string.atof(t[2])

str_porazd = ""
str_cene = ""

for i in db_vsota:
    datum = "%d" % FixDate(db_vsota.index(i))
    str_cene = str_cene + datum
    str_porazd = str_porazd + datum
    for p in db_papirji:
	tecaj = data_cene[p[0]][datumi.index(i[0])]
	vsota = data_vsota[i[0]]
	str_cene = str_cene + " %f" % (tecaj)
	str_porazd = str_porazd + " %f" % (tecaj/vsota * 100)
    str_cene = str_cene + "\n"
    str_porazd = str_porazd + "\n"

MkGraph(str_vsota, "vsota")
MkGraph(str_cene, "cene", "-nxy")
MkGraph(str_porazd, "przd", "-nxy")
